<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <link rel="stylesheet" href="/home/useradmin/Downloads/Lagrange-gh-pages/assets/css/sty.css">
  <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript"></script>
  <!--[if lt IE 9]>
    <script src="//cdnjs.cloudflare.com/ajax/libs/html5shiv/3.7.3/html5shiv-printshiv.min.js"></script>
  <![endif]-->
</head>
<body>
<h3 style="text-align: center;">
pSBVB: Polyploid Sequence Based Virtual Breeding.
</h3>
<p align="center">
A flexible, efficient gene dropping algorithm to simulate sequence based population data and complex traits.
<p align="center">
Miguel Pérez-Enciso
<p align="center">
<p><a href="mailto:miguel.perez@uab.es">miguel.perez@uab.es</a></p>
<p align="center">
<p>With collaborations from N. Forneris, G. de los Campos, A. Legarra and L Zingaretti</p>
<h3 id="purpose">Purpose</h3>
<p>Polyploid sequence based virtual breeding (<strong>pSBVB</strong>) is a modification of <strong>SBVB</strong> software (Pérez-Enciso et al. 2017) that allows simulating traits of an arbitrary genetic complexity in polyploids. Its goal is to simulate complex traits and genotype data starting with a <code>vcf</code> file that contains the genotypes of founder individuals and following a given pedigree. The main output are the genotypes of all individuals in the pedigree and/or molecular relationship matrices (GRM) using all sequence or a series of SNP lists, together with phenotype data. The program implements very efficient algorithms where only the recombination breakpoints for each individual are stored, therefore allowing the simulation of thousands of individuals very quickly. Most of computing time is actually spent in reading the <code>vcf</code> file. Future developments will optimize this step by reading and writing binary mapped files. The <code>vcf</code> file may not contain missing genotypes and is assumed to be phased.</p>
<h3 id="main-features">Main features</h3>
<p><em>Any number of traits. </em>Tool adapted to work with both, auto and allo-polyploid organisms. <em>Any number of QTNs, trait specific. </em>Any number of additive and dominant effects. <em>Can generate a correlation matrix to modelate meiosis in polyploid especies. </em>Can generate correlated allelic effects and frequencies. <em>Efficient algorithms to generate haplotypes and sample SNP genotypes. </em>Computes genomic relationship matrices for any number of SNP arrays simultaneously. <em>It allow to compute Genomic relationship matrix in several ways. </em>Any number of chromosomes, allows for sex chromosomes and varying local recombination rates, that can be sex specific.</p>
<h3 id="installation">Installation</h3>
<p>The source code, manual and examples can be obtained from <a href="https://github.com/mperezenciso/sbvb0" class="uri">https://github.com/mperezenciso/sbvb0</a></p>
<p>To compile:</p>
<p><code>gfortran -O3 kind.f90 ALliball.f90 aux_sub11.f90 psbvb.f90 -o sbvb -lblas</code></p>
<p>or</p>
<p><code>make</code></p>
<p>To install in /usr/local/bin</p>
<p><code>sudo make install</code></p>
<p>The program requires <strong>blas</strong> libraries but these are standard in any unix or OS mac system. We have tested <strong>pSBVB</strong> only in linux with <code>gfortran</code> compiler; intel ifort seems not working, but <code>gfortran</code> in mac OS looks ok.</p>
<h3 id="usage">Usage</h3>
<p>To run (assuming<code>vcf</code> i file is compressed):</p>
<p><code>zcat file.vcf.gz | perl vcf2tped2.pl -hap | cut -d ' ' -f 1,4- | psbvb -isbvb.par</code></p>
<p>Where <code>sbvb.par</code> is the parameter file (details follow). The intermediate steps are simply for <strong>pSBVB</strong> to read genotypes in suitable format, that is,</p>
<p><strong>allele1_snp1_ind1 allele2_snp1_ind1 allele3_snp1_ind1 ... allelep_snp1_ind1 allele1_snp1_ind2 allele2_snp1_ind2 allele3_snp1_ind2 ... allelep_snp1_indp</strong></p>
<p><strong>allele1_snp2_ind1 allele2_snp2_ind1 allele3_snp2_ind1 ... allelep_snp2_ind1 allele1_snp2_ind2 allele2_snp2_ind2 allele3_snp2_ind2 ... allelep_snp2_indp</strong></p>
<p>with alleles coded as <em>0/1</em>. To run the program with the same random seed:</p>
<p><code>… | psbvb –isbvb.par –seed iseed</code></p>
<p>where iseed is an integer number.</p>
<h3 id="parameter-file">Parameter file</h3>
<p>The parameter file controls all <strong>pSBVB</strong> behavior. It consists of a list of sections in UPPER CASE (in any order) followed in the next line by the required data, e.g.,</p>
<p><strong>QTNFILE</strong></p>
<p>sbvb.qtl</p>
<p>tells the program that <strong>QTN</strong> specifications are in sbvb.qtl file. Comments can be mixed starting with # or ! A full list of options in the parameter file is in Appendix 1. In the following, we list the main ones.</p>
<p><strong>PLOIDY</strong></p>
<p>h</p>
<p>tells the program that the organisms used have ploidy <span class="math inline">\(h\)</span>.</p>
<p>Compared to SBVB (designed for diploid organisms), <strong>pSBVB</strong> allows simulating meiosis in autopolyploid or allopolyploid species. For that, <strong>pSBVB</strong> requires a matrix of dimension <span class="math inline">\(h\times h\)</span>, where <span class="math inline">\(h\)</span> is the ploidy level specifying the pairing factors described above. To specify this matrix you must insert in file parameter:</p>
<p><strong>RHOMATRIX</strong></p>
<p>Verrrrrrr como se ponía esta matriz de recombinación</p>
<h3 id="specifying-genetic-architecture">Specifying genetic architecture</h3>
<p>If more than one trait is generated, then use</p>
<p><strong>NTRAIT</strong></p>
<p>ntraits</p>
<p>in parameter file. Otherwise this section is not needed. <strong>pSBVB</strong> requires the user to provide the list of causal SNPs (<strong>QTNs</strong>) as specified in <strong>QTNFILE</strong> section. The format of the QTN file is the next:</p>
<table>
<thead>
<tr class="header">
<th style="text-align: left;">i_chrom</th>
<th style="text-align: center;">i_pos</th>
</tr>
</thead>
<tbody>
</tbody>
</table>
<p>or</p>
<table>
<thead>
<tr class="header">
<th style="text-align: left;">i_chrom</th>
<th style="text-align: center;">i_pos</th>
<th style="text-align: center;">add_eff_Trait_1</th>
<th style="text-align: center;">add_eff_Trait_2</th>
<th style="text-align: center;">...</th>
<th>add_eff_Trait_n</th>
</tr>
</thead>
<tbody>
</tbody>
</table>
<p>or to additive and dominant effects, :blush:</p>
<table style="width:100%;">
<colgroup>
<col style="width: 9%" />
<col style="width: 5%" />
<col style="width: 6%" />
<col style="width: 13%" />
<col style="width: 2%" />
<col style="width: 14%" />
<col style="width: 15%" />
<col style="width: 14%" />
<col style="width: 3%" />
<col style="width: 15%" />
</colgroup>
<thead>
<tr class="header">
<th>i_chrom</th>
<th>i_pos</th>
<th>add_eff_Trait_1</th>
<th style="text-align: left;">add_eff_Trait_2</th>
<th>...</th>
<th>add_eff_Trait_n</th>
<th>dom_eff_Trait_1</th>
<th>dom_eff_Trait_2</th>
<th>...</th>
<th>dom_eff_Trait_n</th>
</tr>
</thead>
<tbody>
</tbody>
</table>
<p><span style="color:green; font-family:Georgia; font-weight: bold; background-color: #FFEF12 ">NOTE:</span> The bellow file, must separated by spaces, and where <em>ichr</em> is chromosome and <em>ipos</em> is position in base pair, <em>add_eff</em> is additive effect, i.e the effect of homozygous alleles and <em>dom_eff</em> is the heterozygous effect.</p>
<p><span style="color:red; font-family:Georgia; font-weight: bold; background-color: #FFFF00 ">WARNING:</span> <strong>QTN</strong> position must coincide with one <strong>SNP</strong> position in the <code>vcf</code> file, otherwise it is not considered.</p>
<p>If QTN effects are not provided, they can be simulated specifying</p>
<p><strong>QTNDISTA</strong></p>
<p>u lower_bound upper_bound | n mu var | g s b</p>
<p>and</p>
<p><strong>QTNDISTD</strong></p>
<p>u lower_bound upper_bound | n mu var | g s b</p>
<p>in parameter file. where 'u' means effects are sampled from a uniform distribution <span class="math inline">\(U\sim(lower_{bound}, upper_{bound})\)</span>, 'n' from a normal distribution <span class="math inline">\(N\sim(mu,var)\)</span> and 'g' from a gamma <span class="math inline">\(\gamma \sim (s,b)\)</span>. For a gamma distribution, you can specify the probability p that a derived allele decreases the phenotype with:</p>
<p><strong>PSIGNQTN</strong></p>
<p>p</p>
<p>The default value is 50%. By default, effects are sampled independently of frequency, i.e., half effects are + and the rest are -, but it is possible to generate a correlation (rho) using the next parameter:</p>
<p><strong>RHOQA</strong></p>
<p>rho</p>
<p>This option can be useful to simulate past selection.</p>
<p>The narrow sense heritability is specified as:</p>
<p><strong>H2</strong></p>
<p>h2</p>
<p>or alternatively, the broad sense heritability (using <strong>H2G</strong>). Only the genotypes from the base population (in the <code>vcf</code> file) are used to adjust heritability.</p>
<h3 id="phenotyping-simulations">Phenotyping simulations</h3>
<p>As <strong>pSBVB</strong> takes ploidy into account to generate the phenotypes and incorporates several options to generate the molecular relationship matrix that are pertinent to polyploids. In a diploid organism, the phenotype for <span class="math inline">\(i\)</span>-th individual can be simulated from:</p>
<p><span class="math inline">\(y_{i}=\mu + \sum_{j=1}^{Q} \gamma_{ij}\alpha_j + \sum_{j=1}^{Q} \delta_{ij}d_j + \epsilon\)</span></p>
<p>Where <span class="math inline">\(\mu\)</span> is the mean general, <span class="math inline">\(\alpha\)</span> is the additive effect of <span class="math inline">\(j\)</span>-th locus, that is, half the expected difference between homozygous genotypes, <span class="math inline">\(\gamma_j\)</span> takes values -1, 0 and 1 for homozygous, heterozygous and alternative homozygous genotypes, respectively. <span class="math inline">\(d_j\)</span> is the dominance effect of <span class="math inline">\(j\)</span>-th locus, and <span class="math inline">\(\delta\)</span> takes value 1 if the genotype is heterozygous, 0 otherwise, and <span class="math inline">\(\epsilon\)</span> is a normal residual. For polyploids, the phenotype of individual <span class="math inline">\(i\)</span> (<span class="math inline">\(y_i\)</span>) (equivalent equation) is simulated from:</p>
<p><span class="math inline">\(y_{i}=\mu + \sum_{j=1}^{Q} \eta_{ij}\alpha_j + \sum_{j=1}^{Q} \phi_{ij}d_j + \epsilon\)</span></p>
<p>where <span class="math inline">\(\eta_j\)</span> is the number of copies of the alternative allele (coded say as 1) minus half the ploidy (<span class="math inline">\(h/2\)</span>) for <span class="math inline">\(j\)</span>-th locus and <span class="math inline">\(i\)</span>-th individual, and <span class="math inline">\(\alpha_j\)</span> is therefore the expected change in phenotype per copy of allele ‘1’ in the <span class="math inline">\(j\)</span>-th locus. In polyploids, as many dominance coefficients as ploidy level (<span class="math inline">\(h\)</span>) minus two can technically be defined. However, this results in an over-parameterized model that is of no practical use. Here instead we define the <span class="math inline">\(\phi_j\)</span> parameter as the minimum number of copies of allele 1 such that the expected phenotype is <span class="math inline">\(d\)</span>. By default, <strong>pSBVB</strong> uses <span class="math inline">\(\phi_j=1\)</span> , that is, any genotype having at least one allele ‘1’ and ‘0’ has the expected phenotypic value <span class="math inline">\(d\)</span>. You can coded <span class="math inline">\(\phi_j\)</span> as any integer between 1 and <span class="math inline">\(h-1\)</span>. Finally, the residual <span class="math inline">\(\epsilon\)</span> is sampled from a <span class="math inline">\(N \sim (0,ve)\)</span>, where <span class="math inline">\(ve\)</span> is adjusted given either <strong>H2</strong> or <strong>H2G</strong> using the genotypes from the base popula2tion. For multiple traits, the fields <strong>H2</strong> or <strong>H2G</strong>, <strong>RHOQA</strong>, and <strong>QTLDISTA</strong> and <strong>QTLDISTD</strong> must be repeated, eg, for two traits:</p>
<p><strong>H2</strong></p>
<p>0.5</p>
<p>0.23</p>
<p><strong>RHOQA</strong></p>
<p>0</p>
<p>-0.4</p>
<p><strong>QTNDISTA</strong></p>
<p>u -0.2 0.2</p>
<p>g 1 0.5</p>
<p>which means that the firts trait have a heredability of <span class="math inline">\(0.5\)</span>, a <strong>RHOQA</strong> parameter of 0 and <strong>QTNDISTA</strong> have an uniform distribution <span class="math inline">\((0.2,0.2)\)</span> and the second trait have a heredability of <span class="math inline">\(0.23\)</span>, <strong>RHOQA</strong> parameter is <span class="math inline">\(-0.4\)</span> and <strong>QTNDISTA</strong> have a gamma distribution with parameters <span class="math inline">\((1,0.5)\)</span></p>
<h3 id="pedigree-file-pedfile">Pedigree file (PEDFILE)</h3>
<p>The format is</p>
<p>id id_father id_mother [sex]</p>
<p>where all ids must be consecutive integers, <span class="math inline">\(0\)</span> if father or mother unknown, sex is optional (<span class="math inline">\(1\)</span> for males, <span class="math inline">\(2\)</span> for females) and only needed if <em>sex</em> chr is specified. The number of individuals in the <code>vcf</code> file must be specified with section:</p>
<p><strong>NBASE</strong></p>
<p>nbase</p>
<p>in the parfile. The pedigree file must contain the first rows as</p>
<table>
<thead>
<tr class="header">
<th><span style="font-weight:normal"> 1 </span></th>
<th><span style="font-weight:normal"> 0 </span></th>
<th><span style="font-weight:normal"> 0 </span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>2</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td>...</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td>nbase</td>
<td>0</td>
<td>0</td>
</tr>
</tbody>
</table>
<p>that is, those in <code>vcf</code> file are assumed to be unrelated.</p>
<h3 id="recombination-map-files">Recombination map files</h3>
<p>By default, <strong>pSBVB</strong> assumes a cM to Mb ratio of 1. This ratio can be changed genomewide with <strong>CM2MB</strong> section in the par file. In addition, local recombination rates can be specified with the <strong>MAPFILE</strong> section. The mapfile takes format</p>
<p><strong>MAPFILE</strong></p>
<table>
<thead>
<tr class="header">
<th style="text-align: center;">ichr</th>
<th style="text-align: center;">last_bp</th>
<th style="text-align: center;">local_cm2mb</th>
</tr>
</thead>
<tbody>
</tbody>
</table>
<p>where <strong>local_cm2mb</strong> is the recombination rate between <strong>last_bp</strong> and previous bound (1 bp if first segment) , or</p>
<table>
<thead>
<tr class="header">
<th style="text-align: center;">ichr</th>
<th style="text-align: center;">last_bp</th>
<th style="text-align: center;">local_cm2mbMales</th>
<th style="text-align: center;">local_cm2mb_females</th>
</tr>
</thead>
<tbody>
</tbody>
</table>
<p>The maximum number of chromosomes allowed by default is 23; should you require more, then section <strong>MAXNCHR</strong> must be included as:</p>
<p><strong>MAXNCHR</strong></p>
<p>nchrom</p>
<p><strong>pSBVB</strong> permits sex chromosomes. The sex chromosome must be declared with <strong>SEXCHR</strong> section. Then, sex 1 is assumed to be the heterogametic sex, and a sex column should be present in the <strong>PEDFILE</strong>.</p>
<p><span style="color:red; font-family:Georgia; font-weight: bold; background-color: #FFFF00 ">WARNING:</span> chromosome ids must be integer consecutive numbers, even for the sex chr if present.</p>
<h3 id="snp-files">SNP files</h3>
<p><strong>pSBVB</strong> can compute the genomic relationship matrix for all sequence data (in two specific ways, see bellow), and/or specific SNP subsets to mimic different genotyping arrays. Several <strong>SNP</strong> lists can be analyzed in the same run repeating the <strong>SNPFILE</strong> section in the par file. Each <strong>SNP</strong> file has the same format as the QTN file, i.e., chromosome and base pair position, as idicated:</p>
<p><strong>SNPFILE</strong></p>
<table>
<thead>
<tr class="header">
<th style="text-align: center;">i_chrom</th>
<th style="text-align: center;">i_pos</th>
</tr>
</thead>
<tbody>
</tbody>
</table>
<p>if you add the command <strong>MIMICDIPLOID</strong> to parameter file, then Genomic relationship matrix is computed assuming than only presence or absence of the alternative allele could be known for the remaining, i.e., although the organism was polyploid, Genomic matrix is computed mimic diploid.</p>
<h2 id="output">Output</h2>
<p>The program writes some general info on the screen, and the following files:</p>
<p>• <strong>OUTYFILE</strong> format (contains phenotypes and breeding values):</p>
<table>
<thead>
<tr class="header">
<th style="text-align: center;"><span class="math inline">\(id\)</span></th>
<th style="text-align: center;"><span class="math inline">\(y\)</span></th>
<th style="text-align: center;"><span class="math inline">\(add_i , i=1,..,ntraits\)</span></th>
<th style="text-align: center;"><span class="math inline">\((add+ dom)_i, i=1,..,ntraits\)</span></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
<p>where <span class="math inline">\(add\)</span> is the first sum in equation of <strong>pSBVB</strong> software, shows above and <span class="math inline">\(dom\)</span> is the second term. For several traits, first are printed all add effects for every trait, next add+dom.</p>
<p>• <strong>OUTQFILE</strong> format (contains <strong>QTN</strong> info):</p>
<table>
<thead>
<tr class="header">
<th style="text-align: center;"><span class="math inline">\(ichr\)</span></th>
<th style="text-align: center;"><span class="math inline">\(pos\)</span></th>
<th style="text-align: center;"><span class="math inline">\(freq_{base}\)</span></th>
<th style="text-align: center;"><span class="math inline">\((add_i)\)</span></th>
<th style="text-align: center;"><span class="math inline">\((dom_i)\)</span></th>
<th style="text-align: center;"><span class="math inline">\(i=1,..,ntraits\)</span></th>
</tr>
</thead>
<tbody>
</tbody>
</table>
<p>where <span class="math inline">\(ichr\)</span> is chromosome, <span class="math inline">\(pos\)</span> is <span class="math inline">\(QTN\)</span> bp position, <span class="math inline">\(freq_{base}\)</span> is frequency in <code>.vcf</code> file, freq is frequency along the pedigree, plus additive, dominant effects and add variance <span class="math inline">\((2pq\alpha^2)\)</span> contribution for each locus by trait.</p>
<p>• <strong>OUTGFILE</strong> format (contains GRM, one per SNPFILE plus sequence) A matrix of <span class="math inline">\(n × n\)</span>, where <span class="math inline">\(n\)</span> is the number of individuals in the pedigree. As many outgfiles as snpfiles are written with subscripts .1, .2 etc. .0 corresponds to sequence. To avoid using sequence, add <strong>NOSEQUENCE</strong> command in parfile.</p>
<p>• <strong>OUTMFILE</strong> format (contains genotypes for evey SNP file and sequence, in plink format optionally using <strong>OUTPLINK</strong> in parfile). As many outmfiles as snpfiles are written with subscripts .1, .2 etc. .0 corresponds to sequence. To avoid using sequence, <strong>NOSEQUENCE</strong> in parfile</p>
<p>Outqfile, outqtn, GRM and marker files are written only if the respective sections <strong>OUTQFILE</strong>, <strong>OUTGFILE</strong> and <strong>OUTMFILE</strong> appear in the <code>.par</code> file. Note in particular that <strong>OUTMFILE</strong> with sequence can be huge! To avoid printing sequence info, use</p>
<p><strong>NOSEQUENCE</strong></p>
<p>in par file.</p>
<p><span style="color:green; font-family:Georgia; font-weight: bold; background-color: #FFEF12 ">NOTE:</span> To compress marker output, include <strong>GZIP</strong> option in parfile.</p>
<h3 id="restart-the-program-keeping-the-same-haplotypes">Restart the program keeping the same haplotypes</h3>
<p>Sometimes one can be interested in running the same experiment but with different genetic architectures or different <strong>SNP</strong> arrays. The program offers two convenient ways to do this as it may keep track of haplotypes so exactly the same genetic structure is preserved, <strong>RESTART</strong> and <strong>RESTARTQTL</strong> options in <code>.par</code> file.</p>
<p>1.With <strong>RESTART</strong>, haplotypes, phenotypes and <strong>QTN</strong> effects are preserved. This is useful to implement selection.</p>
<p>2.With <strong>RESTARTQTN</strong>, haplotypes are preserved but phenotypes and <strong>QTN</strong> effects are sampled again. <strong>RESTARQTN</strong> can be used to run different genetic architectures in the same haplotypes so results can be exactly comparable across models.</p>
<p>The program then writes a <code>.hap</code> file that contains all haplotype structure the first time is run. When <strong>pSBVB</strong> is called again with say another <strong>SNPFILE</strong>, then individuals have the same haplotypes as in previous runs and a new <strong>GRM</strong> can be generated with the new <strong>SNP</strong> file. An important application is to run selection. In fact, <strong>pSBVB</strong> can be run with different pedigree files and the <strong>RESTART</strong> option. <strong>pSBVB</strong> generates only new haplotypes for those individuals not in current <code>.hap</code> file. In a selection scheme, the user should add a new generation pedigree to current pedfile with the offspring of selected individuals. In the new run, <strong>pSBVB</strong> generates haplotypes and phenotypes for the new offspring.</p>
<p><span style="color:#3358ff; font-family:Georgia; font-weight: bold; background-color: #fff933 ">IMPORTANT:</span> The <code>.hap</code> file is used only if <strong>RESTART</strong> is included in parfile. If no <code>.hap</code> file is present, a new one is generated the first time. You can check that <strong>RESTART</strong> is in use checking, e.g, that all phenotypes are the same in different runs.</p>
<p><span style="color:red; font-family:Georgia; font-weight: bold; background-color: #FFFF00 ">WARNING:</span> <strong>RESTARTQTN</strong> is logically not suitable for selection, since effects are sampled anew in each run.</p>
<h3 id="expanding-the-base-population">Expanding the base population</h3>
<p>Very often, complete sequence is available only for very few individuals. <strong>pSBVB</strong> implements an automatic option to generate additional individuals by randomly crossing the available ones and random breeding for a pre specified number of generations. To use this feature, the pedigree file must contain larger number of individuals with unknown parents than in the <code>vcf</code> file. For instance, assume your <code>vcf</code> file contains only four individuals and the pedfile is</p>
<table>
<thead>
<tr class="header">
<th style="text-align: right;"><span style="font-weight:normal"> 1 </span></th>
<th><span style="font-weight:normal"> 0 </span></th>
<th><span style="font-weight:normal"> 0 </span></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">2</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td style="text-align: right;">3</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td style="text-align: right;">...</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="even">
<td style="text-align: right;">20</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="odd">
<td style="text-align: right;">21</td>
<td>1</td>
<td>12</td>
</tr>
<tr class="even">
<td style="text-align: right;">...</td>
<td>...</td>
<td>...</td>
</tr>
</tbody>
</table>
<p>Then individuals 5-20 are generated by randomly crossing 1-4 ids, from id 21 onwards, normal pedigree gene dropping is implemented. The option in parfile is</p>
<p><strong>EXPAND_BASEPOP</strong></p>
<table>
<thead>
<tr class="header">
<th style="text-align: center;">ntgen</th>
<th style="text-align: center;">nfam</th>
</tr>
</thead>
<tbody>
</tbody>
</table>
<p>which means that the new individuals are generated by crossing nfam individuals of the <code>vcf</code> file for ntgen generations.</p>
<h3 id="examples">Examples</h3>
<p>Folder Examples contains two, a toy example consisting in three <strong>SNPs</strong> from a tetraploid individuals and one example consisting of 150 SNPs from the X chromosome of 100 lines from octoploid strawberry lines. The description of the files is:</p>
<p><strong>Base genotypes</strong> test.vcf: original vcf file test.gen: results from cat test.vcf | perl vcf2tped2.pl -hap | cut -d ' ' -f 1,4-</p>
<p>One trait (cat test.gen | sbvb -i test.par) test.par: par fisigmale test.qtn: list of causal SNPs, additive effects are sampled from a gamma test.epi: epi file with two interactions (commented out in test.par by default) test.chip: a list of SNPs from a given array test.outy: phenotype and breeding values test.outq: QTN effects test.outm* : genotypes data test.outg: GRMs</p>
<p>Two traits (cat test.gen | sbvb -i test.par) test2.par test2.qtnsigma test2.epi test2.outq ...</p>
<p><strong>Citation</strong> M. Pérez-Enciso, N. Forneris, G. de los Campos, A. Legarra. An evaluation of sequence-based genomic prediction in pigs using an efficient new simulator. Submitted.</p>
<h2 id="appendix">Appendix</h2>
</body>
</html>
